File: codes/hybrid_hello.f90
program main

  use mpi
  use omp_lib
  implicit none

  integer :: ierr, rank, size, tid, thread_level
  character(len=*), parameter :: FMT1 = "(2(A, I4))"

  call MPI_INIT_THREAD(MPI_THREAD_FUNNELED, thread_level, ierr)
  call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
  call MPI_COMM_SIZE(MPI_COMM_WORLD, size, ierr)

  write(*, FMT1) "Hello, world! I am the main thread of MPI rank ", &
                 rank, " of size ", size

  !$OMP PARALLEL PRIVATE(tid)
  tid = OMP_GET_THREAD_NUM()
  write(*, FMT1) "Hello, world! I am OpenMP thread ", &
                 tid, " of MPI rank ", rank
  !$OMP END PARALLEL

  call MPI_FINALIZE(ierr)

end program main

File: codes/hybrid_hello.c
#include <stdio.h>
#include <omp.h>
#include "mpi.h"

int main(int argc, char* argv[])
{
  int rank, size;
  int tid, thread_level;

  MPI_Init_thread(&argc, &argv, MPI_THREAD_FUNNELED, &thread_level);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

  printf("Hello, world! I am the main thread of MPI rank %d of size %d\n", rank, size);

  #pragma omp parallel private(tid)
  {
    tid = omp_get_thread_num();
    printf("Hello, world! I am OpenMP thread %d of MPI rank %d\n", tid, rank);
  }

  MPI_Finalize();

  return 0;
}
